### PR TITLE
Entitlement plugin api fixes

### PR DESCRIPTION
**Pull request by pierre**
This also fixes logging around `targetDates` in invoice.

### CODE REVIEW

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultBaseEntitlementWithAddOnsSpecifier.java [hunk 1]
```diff
@@ -48,11 +60,19 @@ public UUID getBundleId() {
         return bundleId;
     }
 
+    public void setBundleId(final UUID bundleId) {
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultBaseEntitlementWithAddOnsSpecifier.java]: Mixed reviews... the object was immutable, now it's not. I know it's convenient, but i am wondering if creating another object would not be more maintainable. I am thinking all the payment context for which we have setters, and it's hard to follow the code. if you want to keep it this way, not a strong objection either.

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java [hunk 2]
```diff
@@ -402,7 +402,7 @@ public void uncancelEntitlement(final Iterable<PluginProperty> properties, final
         final WithEntitlementPlugin<Void> uncancelEntitlementWithPlugin = new WithEntitlementPlugin<Void>() {
 
             @Override
-            public Void doCall(final EntitlementApi entitlementApi, final EntitlementContext updatedPluginContext) throws EntitlementApiException {
+            public Void doCall(final EntitlementApi entitlementApi, final DefaultEntitlementContext updatedPluginContext) throws EntitlementApiException {
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java]: Why can't we keep the interface?

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementContext.java [hunk 3]
```diff
@@ -117,6 +119,16 @@ public DefaultEntitlementContext(final OperationType operationType,
         this.tenantId = tenantId;
     }
 
+    private static <T> Iterable<T> merge(final Iterable<T> prevValues, final Iterable<T> newValues) {
```
Comment (sbrossie) [inline entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementContext.java]: I would expect a `merge` method to merge the two input into one - but looks like instead we either return one or the other ? Isn't more like a `lastNonEmptyOrNull` kind of thing?

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultBaseEntitlementWithAddOnsSpecifier.java (hunk 1 above)
Comment (pierre) [reply entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultBaseEntitlementWithAddOnsSpecifier.java]: Unfortunately, all our code is structure around:

```
public interface WithEntitlementPlugin<T> {
    T doCall(final EntitlementApi entitlementApi, final DefaultEntitlementContext updatedPluginContext) throws EntitlementApiException;
}
```

`T` could be `UUID`, `List<UUID>`, etc. So the context has to be mutable (passed in as argument).

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java (hunk 2 above)
Comment (pierre) [reply entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java]: `EntitlementContext` comes from `plugin-api` and doesn't have the setter.

I could have cast'ed, but I prefer being explicit (I almost renamed `DefaultEntitlementContext` as `MutableEntitlementContext` but the changes would be huge).

File: entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementContext.java (hunk 3 above)
Comment (pierre) [reply entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementContext.java]: Indeed - fixed in 30e911ba0ac468d5d9356ce9a5da270d4cadfc9e.
